### ISSUE TITLE
E1142 false & missed alarms

### ISSUE THREAD

**Issue by jakkdl** [pylint-dev/pylint#10074]
### Bug description

```python
def foo():
    yield

async def agen():
    yield

def should_not_give_E1142():
    assert (await x for x in foo())

def should_give_E1142():
    # note the list comprehension
    assert [x async for x in agen()]

# these latter two are correctly handled
def correctly_gives_E1142():
    assert (x for x in await foo())

def correctly_does_not_give_E1142():
    assert (x async for x in agen())
```

See https://github.com/astral-sh/ruff/issues/14167 for the report to ruff that contains a more concrete example.

### Command used

pylint --disable=C pylint_1142.py

### Pylint output

************* Module pylint_1142
pylint_1142.py:8:12: E1142: 'await' should be used within an async function (await-outside-async)
pylint_1142.py:16:23: E1142: 'await' should be used within an async function (await-outside-async)

### Expected behavior

The error on line 16 is correct, but the error on line 8 is a false alarm and it's missing an alarm on line 12.

### Pylint version

$ pylint --version
pylint 3.3.1
astroid 3.3.5
Python 3.12.4 (main, Jun  7 2024, 06:33:07) [GCC 14.1.1 20240522]

### OS / Environment

Arch Linux
